UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Services/FulfilmentCallBackService.cs

```diff
@@ -93,6 +90,67 @@ public async Task<bool> SendCallBackResponse(SalesCatalogueProductResponse sales
             }
         }
 
+        public async Task<bool> SendCallBackErrorResponse(SalesCatalogueProductResponse salesCatalogueProductResponse, SalesCatalogueServiceResponseQueueMessage scsResponseQueueMessage)
+        {
+            salesCatalogueProductResponse.ProductCounts.ReturnedProductCount = 0;
+            salesCatalogueProductResponse.ProductCounts.RequestedProductsNotReturned = new List<RequestedProductsNotReturned> { new RequestedProductsNotReturned { ProductName = null, Reason = essCallBackConfiguration.Value.Reason }};
+
+            if (!string.IsNullOrWhiteSpace(scsResponseQueueMessage.CallbackUri))
+            {
+                try
+                {
+                    ExchangeSetErrorResponse exchangeSetErrorResponse = new ExchangeSetErrorResponse()
+                    {
+                        Links = new ErrorLinks()
+                        {
+                            ExchangeSetBatchStatusUri = new LinkSetBatchStatusUri { Href = $"{fileShareServiceConfig.Value.BaseUrl}/batch/{scsResponseQueueMessage.BatchId}" },
+                            ExchangeSetFileUri = null,
+                            ExchangeSetErrorFileUri = new LinkSetErrorFileUri { Href = $"{fileShareServiceConfig.Value.BaseUrl}/batch/{scsResponseQueueMessage.BatchId}/files/{fileShareServiceConfig.Value.ErrorFileName}" }
+                        },
+                        ExchangeSetUrlExpiryDateTime = Convert.ToDateTime(scsResponseQueueMessage.ExchangeSetUrlExpiryDate).ToUniversalTime(),
+                        RequestedProductCount = salesCatalogueProductResponse.ProductCounts.RequestedProductCount.Value,
+                        ExchangeSetCellCount = salesCatalogueProductResponse.ProductCounts.ReturnedProductCount.Value,
+                        RequestedProductsAlreadyUpToDateCount = salesCatalogueProductResponse.ProductCounts.RequestedProductsAlreadyUpToDateCount.Value,
+                        RequestedProductsNotInExchangeSet = GetRequestedProductsNotInExchangeSet(salesCatalogueProductResponse)
+                    };
+
+                    CallBackErrorResponse callBackErrorResponse = new CallBackErrorResponse()
+                    {
+                        SpecVersion = essCallBackConfiguration.Value.SpecVersion,
+                        Type = essCallBackConfiguration.Value.Type,
+                        Source = essCallBackConfiguration.Value.Source,
+                        Id = Guid.NewGuid().ToString(),
+                        Time = DateTime.UtcNow.ToString("o", CultureInfo.InvariantCulture),
+                        Subject = essCallBackConfiguration.Value.ErrorSubject,
+                        DataContentType = "application/json",
+                        Data = exchangeSetErrorResponse
+                    };
```

Comment:
This code appears to be fairly similar to that in the method above.

Should it be extracted into a separate method/object? (DRY)

Reply:
Resolved, In this piece off code we have actually used different class name and subject value for that reason not able to extracted into a separate method/object.

Reply:
I haven't dived into this to deeply but I'm surprised we cant use the Factory pattern or something similar to return either a ExchangeSetErrorResponse() or ExchangeSetResponse() as both are derived from the same Base class?

Reply:
There is one extra property "ExchangeSetErrorFileUri" in class ExchangeSetErrorResponse which is not there in class ExchangeSetResponse.

Reply:
Resolved ,We have created common class ExchangeSetResponse  and extracted into a separate method.